scipy/stats/_multivariate.py

```diff
@@ -745,11 +745,20 @@ def rvs(self, mean=None, cov=1, size=1, random_state=None):
 
         """
         dim, mean, cov_object = self._process_parameters(mean, cov)
-        cov = cov_object.covariance
-
         random_state = self._get_random_state(random_state)
-        out = random_state.multivariate_normal(mean, cov, size)
-        return _squeeze_output(out)
+
+        if isinstance(cov_object, _covariance.CovViaPSD):
+            cov = cov_object.covariance
+            out = random_state.multivariate_normal(mean, cov, size)
+            out = _squeeze_output(out)
```

Comment:
> to start returning the correct shape to users when they specify a Covariance object explicitly.

Yes

> fix the incorrect shape for CovViaPSD at a later date.

Well, fix the incorrect shape for when the user passes in a regular array at a later day, yes. `CovViaPSD` is not public; it is only used when a user passes in a regular array instead of a `Covariance` object.

> Alternatively, for now, we could continue returning the incorrect shape for all covariance types and fix them (with due warning) later.

Backward incompatible changes are very not fun to work on. They are far more work than actually fixing the code, and then you have to wait a year to see real change.

So my plan now, personally, is to never fix the old classes and just introduce new ones, since there are several other problems (gh-16278). 

> That would keep the behavior across covariance types consistent at all times and avoid the need for branching (which will also be needed in other methods like pdf, logpdf?)

The only branching that needs to happen in the other methods is like:

```python3
if isinstance(cov_object, _covariance.CovViaPSD):
    out = _squeeze_output(out)
```

I'd still branch in this method because the intent was to keep using the same NumPy generator if the user passes in the same regular array. We don't *have* to maintain the same RVS stream for the same seed between feature releases, but I figured we might as well.

> Apologies if this approach was considered and rejected, I couldn't find it in the linked issues.

Yes, well, I considered it. Leaving the squeeze in there for everything is the default, and I went through a little extra trouble to fix the bug when users opt into using the `Covariance` class.

Reply:
That sounds good, thanks for the clarification.